cumulus/client/cirrus-executor/src/worker.rs

```diff
@@ -273,6 +278,31 @@ async fn handle_block_import_notifications<
             number: *header.number(),
         };
 
+        match buffer_sender.feed(()).await {
+            Ok(()) => {
+                if let Err(error) = block_import_throttling_sender.send(()).await {
+                    tracing::error!(
+                        target: LOG_TARGET,
+                        ?error,
+                        "Failed to send the block import advance signal"
+                    );
+                    // Bring down the service as bundles processor is an essential task.
+                    // TODO: more graceful shutdown.
+                    break;
+                }
+            }
+            Err(error) => {
+                tracing::error!(
+                    target: LOG_TARGET,
+                    ?error,
+                    "Failed to reserve a slot in block import throttling buffer"
+                );
+                // Bring down the service as bundles processor is an essential task.
+                // TODO: more graceful shutdown.
+                break;
+            }
+        }
```

Comment:
This construction doesn't seem to be useful. First of all, `feed` will obviously never return an error because receiver lives for at least as long. Second, this construction ties block import of both chains together instead of allowing them to run asynchronously.

This part that reads from `block_imports` needs to run concurrently with `block_imported()` loop, such that you can read a few notifications from the primary chain while notification on the secondary chain is still being processed.